introduce TensorBase::mutable_data_ptr<T>

dagitses opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98781
* #98780
* #98756
* #98755
* #98754
* #98753
* #98752
* #98751
* #98750
* #98734
* #98162
* #98163
* __->__ #97874

See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Commit:
introduce TensorBase::mutable_data_ptr<T>

See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -1554,29 +1554,41 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+   * More efficient helper for Tensor::data_ptr(). Like data<T>(), but
+   * does not do a type check. Unlike the untemplated data(), does
+   * check has_storage() and storage_initialized().
+   */
```
If this is an internal helper, why is it public?

dagitses replied:
Tensor::data_ptr. We can make it private and friend TensorBase if we want.

ezyang replied:
OK. I guess it would have been better if we called this fast or unsafe, but nbd.

dagitses replied:
it's trivial to rename, that can be a separate change altogether.

dagitses commented on aten/src/ATen/templates/TensorMethods.cpp:
```diff
@@ -1,21 +1,36 @@
 namespace at {
 
+// Verifies the requested type is the same as the Tensor's type.
+void check_type(const TensorBase& tensor, ScalarType type, std::string_view type_name) {
```
Yes, but with string_view we know we're checking the length only once.

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Commit:
Update on "introduce TensorBase::mutable_data_ptr<T>"


See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc ezyang bhosmer smessmer ljk53 bdhirsh jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]
